### PR TITLE
Streaming CLI Search

### PR DESCRIPTION
**Pull request by zkat**
This is a WIP `npm search` rewrite branch, based off the work done by @aredridel over at #10615.

The entire search process is now streamed, end-to-end. This branch brings in a couple of stream support libraries to help things along, too.

It should be able to fix pretty much all the heap-blowup crashers in npm.

Fixes: #12619
Fixes: #13247
Fixes: #12587
Fixes: #12439
Fixes: #6254
Fixes: #7090

streamed-search-pretty

It's not great to still be doing a linear search every time, but it's a significant step for search to be at least usable (without crashing or waiting forever for any results).
### TODO
- [x] Stream results from `request`
- [x] Stream results from cache
- [x] Stream output into cache file
- [x] Find issues related to crashing search, close duplicates, and reference master issue here.
- [x] Fix `cb not called` corner case when cache is an empty object (probably when request is empty, too -- check and test both)
- [x] Figure out the `cb not called` thing with `request`'s `res` grrr
- [x] Write unit tests
  - [x] `createEntryStream`
  - [x] `createCacheEntryStream`
  - [x] `createEntryUpdateStream`
  - [x] `createCacheWriteStream`
  - [x] `updateIndex`
- [x] Prettify output
- [x] Do the thing from https://github.com/npm/npm/issues/6254
- [x] Support `--json` for streamed JSON output
- [x] Add message for when there's no search matches
- [X] Rewrite history to not be a crapmess
- [x] check `npm.config.get('unicode')`
- [x] Comb through code and make sure none of the object streams have absurd `highWaterMark` settings.
- [x] Merge and release https://github.com/npm/npm-registry-client/pull/131
- [x] Remove hacked-in copy of `npm-registry-client`
- [x] see if new deps can be minimized (do we need all of `mississippi`? really?) -- decision: keep mississippi cause at least 5 of its utillities are getting used by this code, and it's not worth the extra hassle
- [x] Rewrite integration test for `npm search`
- [x] Document `--tab-separated`
- [x] Refactor all the various stream bits
- [x] Replace `--tab-separated` with `--parseable`
- [x] Make sure searching all fields is well-tested

### CODE REVIEW

**Comment by legodude17** [comment]
This seems really cool. How can I help?

**Comment by zkat** [comment]
@legodude17 Hey! Thanks for the offer -- for now, I think the best thing to do is to just hang in there while I get this PR finished/ready. Since the bulk of the implementation is done, if you find any *glaring* issues with the patches themselves, you can try commenting on them. For the most part, though, a lot about this PR will change before I pass it on to the rest of my team for review.

**Comment by legodude17** [comment]
Okay. Keep up the good work @zkat!

**Comment by timoxley** [comment]
This is awesome. Is the gif in the first post up-to-date?

**Comment by zkat** [comment]
@timoxley yup! The gif is how it looks now.

**Comment by martinheidegger** [comment]
@zkat @legodude17 wouldn't it be good if JSONStream was updated to Streams3 ? https://github.com/dominictarr/JSONStream/pull/61#issuecomment-114876997

**Comment by zkat** [comment]
@martinheidegger Please note that we have to support down to `node@0.10` right now, which rules out relying on `streams3`. Beyond that, there's nothing specific that I'd like our of `streams3` enough to nudge an external project into upgrading.

File: node_modules/sorted-union-stream/index.js
Comment (iarna) [inline node_modules/sorted-union-stream/index.js]: Can you elaborate on what's going on here? And what this being fixed looks like?

File: lib/search.js
Comment (legodude17) [inline lib/search.js]: @zkat, do you think that search sorting will ever be supported?

Comment (zkat) [inline lib/search.js]: @legodude17 not for this iteration, and since this PR is a stopgap to get search working *at all*, it's unlikely that we'll have more featureful searching until we get better backend support for fast searching.

File: lib/search.js
Comment (iarna) [inline lib/search.js]: heehee, a+ emoji use buuut... you need to only include it if `npm.config.get('unicode')` is true.

**Comment by zeg-io** [comment]
Sadly the GIF doesn't work in Safari for some reason :( 
(Works on Chrome though!)

**Comment by legodude17** [comment]
About the last task (That you don't need all of mississippi), based on my initial reading you use almost all of it except concat-stream, from, and to. You could possibly just install those modules (through, stream-finished, pipeline (or whatever they are called)).

File: lib/search.js
Comment (legodude17) [inline lib/search.js]: One option for searching that could work is to stream, but then stream the results to a full object or something, which could then be sorted and displayed. This will decrease the amount of memory required currently, but could still cause a heap blowup if the search term has a lot of results and the computer has a very small amount of RAM.

File: lib/search.js
Comment (othiym23) [inline lib/search.js]: This replaces the setting of `search.usage` immediately above it. That seems unnecessary?

File: lib/search.js
Comment (othiym23) [inline lib/search.js]: I think this callback should be named and hoisted, given its length, complexity, and the fact there are further callbacks nested in it.

**Review by othiym23 — COMMENTED** [review]
This is looking good! I think I'm going to do another pass through this at the individual commit level, because there are too many files in the PR to use the global view (maybe after you do an initial squash?). I also think I'll probably want to pull the branch down and play with it / try to break it some before it gets squashed and merged. It might be worth putting all of the stuff that affects both result and cache streaming into a single file, pulled out of `lib/cache/update-index.js` and `lib/search.js`, because right now it's a little hard to correlate some of the stream hijinx between the two. Aside from that, the architectural basis looks sound.

I do have some concerns about the amount of dependencies we're adding with this change, though. Are we using enough of `mississippi` to justify using the whole package, or might we save some space by only including the bits that we're using?

Again, this is looking good, and I appreciate all of the care and time you've put into this. This is a huge improvement over what we have now.

File: doc/cli/npm-search.md
Comment (zkat) [inline doc/cli/npm-search.md]: I remember we had an over-the-air conversation about `--parseable` vs `--tab-separated` but I don't remember if there was a particular reason for it. So then: would you like me to use `--parseable` instead for this? It's no skin off my back. I have no strong opinion either way, and I understand your point. (and, fwiw, being specific about what the "parseable" format is seems a little nicer, but it's not the convention we already have)

**Comment by zkat** [comment]
@othiym23 I've addressed all the specific comments now, I think, and now to respond to the initial review comment:

> It might be worth putting all of the stuff that affects both result and cache streaming into a single file, pulled out of `lib/cache/update-index.js` and `lib/search.js`.

I'm not entirely sure how to parse this one: the divide between those two files is relatively clear to me. The API of `update-index.js` is pretty much to give you a stream, by whatever means it feels appropriate, of all package entries. As its own internal side-effect, it does caching, but the cache is not something `search.js` is really aware of at all. Did I miss something here about what affects both of them, that you're referring to? I don't mind moving around whatever, I'm just not sure which things. :)

> I do have some concerns about the amount of dependencies we're adding with this change, though. Are we using enough of mississippi to justify using the whole package, or might we save some space by only including the bits that we're using?

I was a bit concerned, too, but the only significant dependency we don't use is `concat-stream` (376K uncompressed, 60K tarballed up). The others are tiny (`flush-write-stream`: 28K uncompressed; `stream-each`: 32K uncompressed), and I can totally see us using all 3 eventually. All other utilities are used either by the code, or by the tests. I started to install all the mississippi deps individually at one point, and it just made things a lot noisier, and I realized I was dragging a significant % of the library into a bunch of files. I think it's nice to have it bundled like this. I can go ahead and do that anyway if you're uncomfortable with the 60K+ that the unused deps will add, though.

**Comment by othiym23** [comment]
> I'm not entirely sure how to parse this one: the divide between those two files is relatively clear to me.

Both are working with streams to a single representation on disk. There isn't really a datatype for that representation, so there's not any kind of encapsulated interface, which makes me concerned that a. it will be tough for newbies (as it was for me, a lil bit) to correlate the two different halves of updating that cache and scanning it, and b. that it's easier for the two chunks of code to drift if we're not careful (especially via contributors making patches or changes – see a.). From looking at the code, it _seems_ like all the streaming code can be pulled into a single module with a defined interface, but it could be that it's tricky in practice. What do you think?

>  I can go ahead and do that anyway if you're uncomfortable with the 60K+ that the unused deps will add, though.

The npm package tarball is now over 3MiB, which feels large to me. However, the point about noise is a good one, and there is a good chance we'll end up using those things later. Consider that case made. I just want to encourage a healthy sense of parsimony on the part of the team, because I think we haven't really been paying attention to how large the CLI is getting, and it has a significant effect on everyone downstream from us.

Still waiting on the squash for a complete, final review!

File: doc/cli/npm-search.md
Comment (othiym23) [inline doc/cli/npm-search.md]: Yeah, I think I would. It's fine if it stays tab-separated, but I don't think we need a separate format for tab-separated results just for this subcommand. Just out of curiosity, what happens if there are tabs embedded in the description?

File: test/tap/update-index.js
Comment (othiym23) [inline test/tap/update-index.js]: Is this a matter of it being tricky to get the configuration here? Or having control over what times things happen at, and avoiding race conditions? This particular TODO has a bit of a smell for me.

File: test/tap/search.js
Comment (othiym23) [inline test/tap/search.js]: Do any of the tests in here match against the description field? Or username fields? These are both things that I've found surprisingly useful in the past, when search worked. I'm pretty sure they work in practice (testing right now), but it would be good to have them included in the tests.

**Review by othiym23 — COMMENTED** [review]
Thanks for the rebase! I think I've checked everything I wanted here, and replied to your responses from the last go-round. This is looking good! I wouldn't consider any of the items I've raised in this round blockers, but several of them are changes I think we'd be happy we made later. I'm still trying to exercise edge cases a little more, and will let you know if I find anything egregious (although the tests, which are great, do a pretty good job of nailing down the edges). ガンバレよー！

File: lib/cache/update-index.js
Comment (zkat) [inline lib/cache/update-index.js]: This is a good idea!

Comment (zkat) [inline lib/cache/update-index.js]: done!

File: lib/cache/update-index.js
Comment (zkat) [inline lib/cache/update-index.js]: This is done.

File: lib/search.js
Comment (zkat) [inline lib/search.js]: This was done as part of the refactor. All these options are passed in the `opts` object to the formatter stream.

File: test/tap/update-index.js
Comment (zkat) [inline test/tap/update-index.js]: This comment predates a lot of stuff I ended up doing to make doubleplus sure these timestamps were good in the tests. I've removed it.

File: lib/search.js
Comment (othiym23) [inline lib/search.js]: Changing this to include the description flag is in keeping with the spirit of keeping the config out of the refactored search, but the cascading effects of this are pretty big. It's also weird that there are two options arguments here now.

File: lib/search/package-filter.js
Comment (othiym23) [inline lib/search/package-filter.js]: Picking up from up above, this is a hella confusing arguments list: should it be `included`, `excluded`, and `opts`?

File: lib/search.js
Comment (othiym23) [inline lib/search.js]: Looking at this in the context of `filter()`, fixing this could be as simple as changing `opts` to `included`, and `searchexclude` to `excluded` to match.

File: lib/search/all-package-metadata.js
Comment (othiym23) [inline lib/search/all-package-metadata.js]: Outstanding. :tada:

File: lib/search/all-package-metadata.js
Comment (othiym23) [inline lib/search/all-package-metadata.js]: Good eye, making this match with the above.

File: doc/misc/npm-config.md
Comment (othiym23) [inline doc/misc/npm-config.md]: Good start. This is yet more evidence that [`doc/misc/config.md` must be destroyed](https://github.com/npm/npm/issues/7185), because now I really want this to include what the parseable form means for other commands (which is totally out of scope for this PR – DON'T DO THIS).

File: lib/search.js
Comment (othiym23) [inline lib/search.js]: _sings the "this is why I like tests" song_

File: lib/search/format-package-stream.js
Comment (othiym23) [inline lib/search/format-package-stream.js]: Functions feel less lonely with names (also, the CLI's code base has a house style of using

```js
module.exports = hoistedFunctionName
function hoistedFunctionName () {
//...
```
)

File: lib/search/all-package-metadata.js
Comment (othiym23) [inline lib/search/all-package-metadata.js]: Same comment as below giving this function a name. Otherwise, _so beautiful_ 😭.

File: lib/search/stream-filter.js
Comment (othiym23) [inline lib/search/stream-filter.js]: Does this need to be a standalone file, given that it's only used in one place and is a pretty simple direct invocation?

**Review by othiym23 — APPROVED** [review]
Found a bunch of small things that you can tweak or not, to your taste. I don't think this will need another round of review, although it would probably be good to do another pass at squashing commits (all of the post-refactor test changes can probably be squashed into the refactor changesets, for example).

Your refactor is pretty much exactly what I had in mind. Thank you for bearing with me on that! I think the result is a lot clearer than what we had before, and I think should make things simpler when we decide to make further changes to search.

The documentation for https://docs.npmjs.com/misc/config#searchopts and https://docs.npmjs.com/misc/config#searchexclude are pretty terrible, in that I had to read the code to learn exactly what they did. It might be good to update `doc/misc/config.md` to include a bit more detail about how the two are meant to be used (including mentioning that they're treated like JS regexps, which is a pretty cool feature that I bet _almost nobody_ who hasn't been using npm since 2011 or earlier knows about).

Congratulations! This is really good work!

**Comment by othiym23** [comment]
We _do_ have to get the tests passing before `npm@4` is released, though. I know that @iarna was looking at the Travis failures – maybe the two of you could pair and see if you can get everything ironed out?

File: lib/search/stream-filter.js
Comment (zkat) [inline lib/search/stream-filter.js]: This is such a baseline tool that I wanted to hide the implementation but it wasn't worth adding `stream-filter` as a dependency. I guess I can just define it as a utility in `search.js`

**Pull request merged**
